stream/stream_smb2.c

```diff
@@ -0,0 +1,440 @@
+    p->share = malloc(share_length + 1);
+    strncpy(p->share, share_start + 1, share_length);
+    *(p->share + share_length) = '\0';
+    MP_DBG(stream, "[smb] share: %s\n", p->share);
+
+    //split by @
+    char *server_start = strstr(author_str, "@");
+    const size_t server_length = strlen(author_str) -
+        (server_start - author_str) - 1;
+    //printf("server_length: %zd\n", server_length);
+    p->server = malloc(server_length + 1);
+    strncpy(p->server, server_start + 1, server_length);
+    *(p->server + server_length) = '\0';
+    MP_DBG(stream, "[smb] server: %s\n", p->server);
+
+    author_str[server_start - author_str] = '\0';
+    if (strlen(author_str) < 3) {//handle smb://:@server/share/fielpath
+        MP_WARN(stream, "[smb] handle special '%s'\n", author_str);
+        free(author_str);
+        return 0;
+    }
+    char **tokens = str_split(author_str, ':');
+    size_t tokens_length = 0;
+    if (tokens) {
+        size_t i = 0;
+        while (*(tokens + i)) {
+            i++;
+        }
+        tokens_length = i;
+    }
+    if (tokens_length == 1) {
+        //only user
+        const size_t user_length = strlen(*tokens);
+        p->user = malloc(user_length + 1);
+        strncpy(p->user, *tokens, server_length);
+        *(p->user + user_length)='\0';
+        MP_DBG(stream, "[smb] user: '%s'\n", p->user);
+    } else if (tokens_length == 2) {
+        //user:password
+        const size_t user_length = strlen(*tokens);
+        p->user = malloc(user_length + 1);
+        strncpy(p->user, *tokens, server_length);
+        *(p->user + user_length)='\0';
+        MP_DBG(stream, "[smb] user: '%s'\n", p->user);
+
+        const size_t password_length = strlen(*(tokens + 1));
+        p->password = malloc(password_length + 1);
+        strncpy(p->password, *(tokens + 1), password_length);
+        *(p->password + password_length)='\0';
+        MP_DBG(stream, "[smb] password: '%s'\n", p->password);
+    } else if (tokens_length == 3) {
+        //domain:user:password
+        const size_t domain_length = strlen(*tokens);
+        p->domain = malloc(domain_length + 1);
+        strncpy(p->domain, *tokens, domain_length);
+        *(p->domain + domain_length)='\0';
+        MP_DBG(stream, "[smb] domain: %s\n", p->domain);
+
+        const size_t user_length = strlen(*(tokens + 1));
+        p->user = malloc(user_length + 1);
+        strncpy(p->user, *(tokens + 1), user_length);
+        *(p->user + user_length)='\0';
+        MP_DBG(stream, "[smb] user: %s\n", p->user);
+
+        const size_t password_length = strlen(*(tokens + 2));
+        p->password = malloc(password_length + 2);
+        strncpy(p->password, *(tokens + 2), password_length);
+        MP_DBG(stream, "[smb] password: %s\n", p->password);
+        *(p->password + password_length)='\0';
+    }
+    free(author_str);
+    if ((p->server) != NULL && (p->share) != NULL && (p->filepath) != NULL) {
+        return 0;
+    }
+    return -1;
+}
```

Comment:
If you use mallocs, you are required to check for allocation failure. You can do this in a trivial way, like using MP_HANDLE_OOM (which performs a controlled crash on allocation failures, rather than triggering UB).

But in general, you should probably use talloc. Also, the bstr API makes string parsing much easier. Last but not least, I wonder if the SMB libraries does not provide helpers for this, or whether things like passwords shouldn't be separate options.